### PR TITLE
Address compiler warnings

### PR DESCRIPTION
**Pull request by damieng**
Mostly just around null annotations but also a few unused properties etc.

Solution now compiles 100% clean with Roslyn, no warnings.

### CODE REVIEW

File: tests/MongoDB.EntityFrameworkCore.FunctionalTests/MongoDB.EntityFrameworkCore.FunctionalTests.csproj [hunk 1]
```diff
@@ -1,30 +1,30 @@
...
-        <ImplicitUsings>enable</ImplicitUsings>
-        <Nullable>enable</Nullable>
+  <PropertyGroup>
+    <TargetFramework>net8.0</TargetFramework>
+    <ImplicitUsings>enable</ImplicitUsings>
+    <Nullable>enable</Nullable>
+    <IsPackable>false</IsPackable>
+    <IsTestProject>true</IsTestProject>
+    <NoWarn>EF1001,CS8602,CS8618</NoWarn>
+  </PropertyGroup>
```
Comment (BorisDog) [inline tests/MongoDB.EntityFrameworkCore.FunctionalTests/MongoDB.EntityFrameworkCore.FunctionalTests.csproj:10]: Does it makes sense to migrate those settings (or a subset) to` tests/Directory.Build.props` ?

Comment (damieng) [reply tests/MongoDB.EntityFrameworkCore.FunctionalTests/MongoDB.EntityFrameworkCore.FunctionalTests.csproj:10]: I don't think it's worth it yet. There won't likely be more test projects or any new stuff any time soon.

Comment (BorisDog) [reply tests/MongoDB.EntityFrameworkCore.FunctionalTests/MongoDB.EntityFrameworkCore.FunctionalTests.csproj:10]: So we have `Directory.Build.props` already, seems like a natural place for such settings.

**Review by BorisDog — APPROVED** [review]
LGTM

**Pull request merged**